extensions/web/lib.rs

```diff
@@ -15,12 +34,197 @@ pub fn init() -> Extension {
+) -> Result<String, AnyError> {
+  let cfg = base64::Config::new(base64::CharacterSet::Standard, true)
+    .decode_allow_trailing_bits(true);
+  let out = base64::encode_config(&s, cfg);
+  Ok(out)
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DecoderOptions {
+  label: String,
+  ignore_bom: bool,
+  fatal: bool,
+}
+
+fn op_encoding_normalize_label(
+  _state: &mut OpState,
+  label: String,
+  _: (),
+) -> Result<String, AnyError> {
+  let encoding = Encoding::for_label(label.as_bytes()).ok_or_else(|| {
+    range_error(format!(
+      "The encoding label provided ('{}') is invalid.",
+      label
+    ))
+  })?;
+  Ok(encoding.name().to_lowercase())
+}
+
+fn op_encoding_new_decoder(
+  state: &mut OpState,
+  options: DecoderOptions,
+  _: (),
+) -> Result<ResourceId, AnyError> {
+  let DecoderOptions {
+    label,
+    fatal,
+    ignore_bom,
+  } = options;
+
+  let encoding = Encoding::for_label(label.as_bytes()).ok_or_else(|| {
+    range_error(format!(
+      "The encoding label provided ('{}') is invalid.",
+      label
+    ))
+  })?;
+
+  let decoder = if ignore_bom {
+    encoding.new_decoder_without_bom_handling()
+  } else {
+    encoding.new_decoder_with_bom_removal()
+  };
+
+  let rid = state.resource_table.add(TextDecoderResource {
+    decoder: RefCell::new(decoder),
+    fatal,
+  });
+
+  Ok(rid)
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct DecodeOptions {
+  rid: ResourceId,
+  stream: bool,
+}
+
+fn op_encoding_decode(
+  state: &mut OpState,
+  data: ZeroCopyBuf,
+  options: DecodeOptions,
+) -> Result<String, AnyError> {
+  let DecodeOptions { rid, stream } = options;
+
+  let resource = state
+    .resource_table
+    .get::<TextDecoderResource>(rid)
+    .ok_or_else(bad_resource_id)?;
+
+  let mut decoder = resource.decoder.borrow_mut();
+  let fatal = resource.fatal;
+
+  let max_buffer_length = if fatal {
+    decoder
+      .max_utf8_buffer_length_without_replacement(data.len())
+      .ok_or_else(|| range_error("Value too large to decode."))?
+  } else {
+    decoder
+      .max_utf8_buffer_length(data.len())
+      .ok_or_else(|| range_error("Value too large to decode."))?
+  };
+
+  let mut output = String::with_capacity(max_buffer_length);
+
+  if fatal {
+    let (result, _) =
+      decoder.decode_to_string_without_replacement(&data, &mut output, !stream);
+    match result {
+      DecoderResult::InputEmpty => Ok(output),
+      DecoderResult::OutputFull => {
+        Err(range_error("Provided buffer too small."))
+      }
+      DecoderResult::Malformed(_, _) => {
+        Err(type_error("The encoded data is not valid."))
+      }
+    }
+  } else {
+    let (result, _, _) = decoder.decode_to_string(&data, &mut output, !stream);
+    match result {
+      CoderResult::InputEmpty => Ok(output),
+      CoderResult::OutputFull => Err(range_error("Provided buffer too small.")),
+    }
+  }
+}
+
+struct TextDecoderResource {
+  decoder: RefCell<Decoder>,
```

Comment:
Is this `RefCell<>` needed? Seems superfluous

Reply:
Yes, because I need mutable access to the decoder, and I can't do that with Rc alone.

Reply:
Right, non-issue then